### DISCUSSION TITLE
float-compare: questions with best practices

### DISCUSSION THREAD

**Discussion by mmorel-35** [Antonboom/testifylint#151] [Q&A]
In https://github.com/argoproj/argo-cd/pull/18721/files I applyied float-compare rules .
I have two questions here.
With this first case we have situation 
```go
assert.Equal(t, float64(1), retryCount)
```
This suggestion 
```go
assert.InEpsilon(t, float64(1), retryCount, 0.0001)
```
Having to convert an integer to a float to compare with a float then have to be used with `assert.InDelta` or `assert.InEpsilon`.

What do you think about using 
 ```go
assert.EqualValue(t, 1, retryCount)
```
Maybe this situation indicates a mistyping at the origin so the solution is to change the type of retryCount as a counter is usually an integer.

The other question is about 
```go
assert.Equal(t, liveReplicas, desiredReplicas)
```
I was suggested to use `assert.InEpsilon` but I ended up using

```go
assert.Zero(t, desiredReplicas)
assert.Zero(t, liveReplicas)
```

The reason behind this is that `assert.InEpsilon` doesn’t support having 0 as an expected value. The tricky thing in this situation is that desiredReplicas might not be known in certain cases.

**Comment by Antonboom** [comment]
> Maybe this situation indicates a mistyping at the origin so the solution is to change the type of retryCount as a counter is usually an integer.

Nice point, looks like a feature 👍 

> The reason behind this is that assert.InEpsilon doesn’t support having 0 as an expected value.

Why not `assert.InDelta(t, 0., 0., 0.)`?

**Comment by mmorel-35** [comment]
I don't see why InDelta would support 0 as an expected value while InEpsilon doesn't. It didn't  occured to me to even try it .

**Comment by Antonboom** [comment]
@mmorel-35, I want to transfer this issue into discussion.
But looks like I have no right for it, wtf? )

**Comment by ccoVeille** [comment]
I also have a question with float compared to 0

Something was raised I a recent review when @mmorel-35 used testifylint + manual edit to fix a code using somehow

```go
var f float
assert.Equal(t, 0, f)
```

**Reply by ccoVeille** [reply]
Here exactly

https://github.com/prometheus/prometheus/pull/14334#discussion_r1649615344

**Reply by ccoVeille** [reply]
I'm wondering why using Zero(the, f) would be valid compared to Equal(t, 0, f)

And why in epsilon or in delta won't accept 0, @mmorel-35 point here.

I mean either in don't get something.

Or we have something to detect with zero used with float

And if you are wondering @Antonboom, I didn't look at the code yet.

I'm raising the point as a note for the discussion, either for myself and for you guys
